### PR TITLE
[LZ4HC] Speed up pattern compression with external dictionary

### LINKED ISSUE #761: [LZ4HC] Speed up pattern compression with external dictionary
LZ4HC detects patterns in the source data, like a huge run of 0s, and instead of becoming an N^2 algorithm, special cases it. However, this special case is only applied to matches in the current segment. Matches into the external dictionary skip this optimization. See https://github.com/lz4/lz4/blob/dev/lib/lz4hc.c#L317.

This task is to extend this optimization to apply to matches in the external dictionary.

### PR DESCRIPTION
**Pull request by terrelln**
Fixes #761.

Tested by running the ossfuzz round trip fuzzers overnight without any errors (but they caught bugs during development and in the initial version quickly).

### CODE REVIEW

File: lib/lz4hc.c [hunk 1]
```diff
@@ -313,34 +326,41 @@ LZ4HC_InsertAndGetWiderMatch (
...
+                        const BYTE* const dictStart = dictBase + hc4->lowLimit;
+                        const BYTE* const iLimit = extDict ? dictBase + dictLimit : iHighLimit;
+                        size_t forwardPatternLength = LZ4HC_countPattern(matchPtr+sizeof(pattern), iLimit, pattern) + sizeof(pattern);
+                        if (extDict && ip + forwardPatternLength == iLimit) {
+                            U32 const rotatedPattern = LZ4HC_rotatePattern(forwardPatternLength, pattern);
+                            forwardPatternLength += LZ4HC_countPattern(lowPrefixPtr, iHighLimit, rotatedPattern);
+                        }
+                        {   const BYTE* const lowestMatchPtr = extDict ? dictStart : lowPrefixPtr;
+                            size_t const backLengthRaw = LZ4HC_reverseCountPattern(matchPtr, lowestMatchPtr, pattern);
+                            size_t const backLength = matchCandidateIdx - MAX(matchCandidateIdx - (U32)backLengthRaw, lowestMatchIndex);
```
Comment (terrelln) [inline lib/lz4hc.c]: This is the last changed line, after this everything is the same, just indented one more level.

**Comment by Cyan4973** [comment]
Is there some way to measure the speed improvement ?

**Comment by terrelln** [comment]
I'll measure the speed on zeros and normal data

**Comment by terrelln** [comment]
```
make clean round_trip_stream_fuzzer DEBUGLEVEL=0
time ./round_trip_stream_fuzzer clusterfuzz-testcase-minimized-round_trip_stream_fuzzer-5658930090344448
```

This file (attached) takes 9 seconds to compress before the patch and 0.1 seconds to compress after.

[clusterfuzz-testcase-minimized-round_trip_stream_fuzzer-5658930090344448.zip](https://github.com/lz4/lz4/files/3428652/clusterfuzz-testcase-minimized-round_trip_stream_fuzzer-5658930090344448.zip)

Also the following command takes 6 seconds on dev and 0.01 seconds after this patch.

```
time head -c 8M /dev/zero | ./lz4 -BD -12 > /dev/null
```

I don't see a noticeable speed difference when running the benchmark tool at level 12 with and without `-BD` before and after the patch.

**Comment by Cyan4973** [comment]
Excellent, thanks @terrelln !

**Pull request merged**